### ISSUE TITLE
Can not run Distributed Training with RDMA

### ISSUE THREAD

**Issue by mengkai94** [bytedance/byteps#216]
**Describe the bug**
I use byteps  according to Distributed Training with RDMA of byteps/docs/step-by-step-tutorial.
I use the latest images : bytepsimage/tensorflow and get error.
But using bytepsimage/tensorflow_rdma and bytepsimage/server_rdma can success.
Using one scheduler, one server, two workers with one respective gpu in a physical worker.

**To Reproduce**

1. For the scheduler:
docker run -it --net=host --device /dev/infiniband/rdma_cm --device /dev/infiniband/issm0 --device /dev/infiniband/ucm0 --device /dev/infiniband/umad0 --device /dev/infiniband/uverbs0 --cap-add IPC_LOCK byteps:tensorflow-0.2 bash
export DMLC_ENABLE_RDMA=1
export DMLC_NUM_WORKER=2
export DMLC_ROLE=scheduler
export DMLC_NUM_SERVER=1
export DMLC_INTERFACE=eth0
export DMLC_PS_ROOT_URI=xxx.xx.xx.xx
export DMLC_PS_ROOT_PORT=9008
bpslaunch

2. For the server:
docker run -it --net=host --device /dev/infiniband/rdma_cm --device /dev/infiniband/issm0 --device /dev/infiniband/ucm0 --device /dev/infiniband/umad0 --device /dev/infiniband/uverbs0 --cap-add IPC_LOCK byteps:tensorflow-0.2 bash
export DMLC_ENABLE_RDMA=1
export DMLC_NUM_WORKER=2
export DMLC_ROLE=server
export DMLC_NUM_SERVER=1
export DMLC_INTERFACE=eth0
export DMLC_PS_ROOT_URI=xxx.xx.xx.xx
export DMLC_PS_ROOT_PORT=9008
bpslaunch

3. For worker-0:
nvidia-docker run -it --net=host --shm-size=32768m --device /dev/infiniband/rdma_cm --device /dev/infiniband/issm0 --device /dev/infiniband/ucm0 --device /dev/infiniband/umad0 --device /dev/infiniband/uverbs0 --cap-add IPC_LOCK  byteps:tensorflow-0.2 bash
export NVIDIA_VISIBLE_DEVICES=0
export DMLC_ENABLE_RDMA=1
export DMLC_WORKER_ID=0 
export DMLC_NUM_WORKER=2
export DMLC_ROLE=worker
export DMLC_NUM_SERVER=1
export DMLC_INTERFACE=eth0
export DMLC_PS_ROOT_URI=xxx.xx.xx.xx
export DMLC_PS_ROOT_PORT=9008
bpslaunch python3 /usr/local/byteps/example/tensorflow/synthetic_benchmark.py --model ResNet50 --num-iters 1000000 

4. For worker-1:
nvidia-docker run -it --net=host --shm-size=32768m --device /dev/infiniband/rdma_cm --device /dev/infiniband/issm0 --device /dev/infiniband/ucm0 --device /dev/infiniband/umad0 --device /dev/infiniband/uverbs0 --cap-add IPC_LOCK byteps:tensorflow-0.2 bash
export NVIDIA_VISIBLE_DEVICES=7
export DMLC_ENABLE_RDMA=1
export DMLC_WORKER_ID=1 
export DMLC_NUM_WORKER=2
export DMLC_ROLE=worker
export DMLC_NUM_SERVER=1
export DMLC_INTERFACE=eth0
export DMLC_PS_ROOT_URI=xxx.xx.xx.xx
export DMLC_PS_ROOT_PORT=9008
bpslaunch python3 /usr/local/byteps/example/tensorflow/synthetic_benchmark.py --model ResNet50 --num-iters 1000000 

5. scheduler see error
BytePS launching scheduler
[02:34:49] byteps/server/server.cc:339: BytePS server engine uses 4 threads, consider increasing BYTEPS_SERVER_ENGINE_THREAD for higher performance
[02:34:49] src/postoffice.cc:20: enable RDMA for networking
[02:34:49] src/./rdma_van.h:40: Shared memory IPC has been disabled
[02:34:49] src/./rdma_van.h:801: OnConnect to Node 1 with Transport=RDMA
[02:34:49] src/./rdma_van.h:207: Connect to Node 1 with Transport=RDMA
[02:34:58] src/./rdma_van.h:801: OnConnect to Node 2147483647 with Transport=RDMA
[02:35:23] src/./rdma_van.h:801: OnConnect to Node 2147483647 with Transport=RDMA
[02:35:38] src/./rdma_van.h:801: OnConnect to Node 2147483647 with Transport=RDMA
[02:35:38] src/./rdma_van.h:207: Connect to Node 9 with Transport=RDMA
[02:35:38] src/./rdma_van.h:207: Connect to Node 8 with Transport=RDMA
[02:35:38] 3rdparty/ps-lite/include/dmlc/logging.h:276: [02:35:38] src/./rdma_transport.h:130: Check failed: mr
Stack trace returned 7 entries:
[bt] (0) /usr/local/lib/python3.6/dist-packages/byteps-0.2.0-py3.6-linux-x86_64.egg/byteps/server/c_lib.cpython-36m-x86_64-linux-gnu.so(+0x1b98c) [0x7f788672398c]
[bt] (1) /usr/local/lib/python3.6/dist-packages/byteps-0.2.0-py3.6-linux-x86_64.egg/byteps/server/c_lib.cpython-36m-x86_64-linux-gnu.so(+0x1bdad) [0x7f7886723dad]
[bt] (2) /usr/local/lib/python3.6/dist-packages/byteps-0.2.0-py3.6-linux-x86_64.egg/byteps/server/c_lib.cpython-36m-x86_64-linux-gnu.so(+0x40fb8) [0x7f7886748fb8]
[bt] (3) /usr/local/lib/python3.6/dist-packages/byteps-0.2.0-py3.6-linux-x86_64.egg/byteps/server/c_lib.cpython-36m-x86_64-linux-gnu.so(+0x57dbb) [0x7f788675fdbb]
[bt] (4) /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0xbd66f) [0x7f7885e0866f]
[bt] (5) /lib/x86_64-linux-gnu/libpthread.so.0(+0x76db) [0x7f78891356db]
[bt] (6) /lib/x86_64-linux-gnu/libc.so.6(clone+0x3f) [0x7f788946e88f]

**Expected behavior**
run sucess

**Screenshots**
scheduler
![image](https://user-images.githubusercontent.com/27659870/76178850-be8bc680-61f3-11ea-8d7f-50d2b41f06d2.png)
server
![image](https://user-images.githubusercontent.com/27659870/76178857-c6e40180-61f3-11ea-9fc0-d26433876a04.png)
worker0
![image](https://user-images.githubusercontent.com/27659870/76178876-dc592b80-61f3-11ea-9882-308b9d48444f.png)
worker1
![image](https://user-images.githubusercontent.com/27659870/76178885-e24f0c80-61f3-11ea-8958-f9eacafdf3d3.png)

**Environment physical machine(please complete the following information):**
 - OS:16.04.2-Ubuntu
 - GCC version: 5.4.0
 - CUDA and NCCL version:10.1
 - Framework (TF, PyTorch, MXNet):TF

**Comment by ymjiang** [comment]
Seems like it hits the limit of memory registration. We will post a fix soon. Before that,  could you please manually change the value of `kReplyDepth` (in byteps/3rdparty/ps-lite/src/rdma_utils.h) to 256 (or even smaller), and then recompile with the following commands?

```
cd byteps/3rdparty/ps-lite && make clean  && cd -
pip3 uninstall -y byteps
python3 setup.py clean --all
python3 setup.py install
```

**Comment by mengkai94** [comment]
It works! thanks you, @ymjiang .
Does the value of  kReplyDepth effect byteps's performance？

**Comment by bobzhuyb** [comment]
@mengkai94 256 should be enough in most cases.

**Comment by mengkai94** [comment]
thanks you.

**Comment by mengkai94** [comment]
Using same setting in k8s does not work. Scheduler/server/workers are in same physical machine.
Workers do not start run, but using bytepsimage/tensorflow_rdma and bytepsimage/server_rdma  in k8s can success. 
BTW, does byteps need ssh_port same with DMLC_PS_ROOT_PORT?

#Screenshots
scheduler
![image](https://user-images.githubusercontent.com/27659870/76189623-218e5500-6216-11ea-9567-013c3e16b98c.png)
server
![image](https://user-images.githubusercontent.com/27659870/76189651-34a12500-6216-11ea-8b65-2047b2f2f563.png)
woker0
![image](https://user-images.githubusercontent.com/27659870/76189667-3cf96000-6216-11ea-8534-35a8c30ca903.png)
worker1
![image](https://user-images.githubusercontent.com/27659870/76189689-4b477c00-6216-11ea-85a1-4ca6c819fd40.png)

**Comment by bobzhuyb** [comment]
DMLC_PS_ROOT_PORT does not need to be the same as SSH port. Instead, it should be a different port otherwise the it will conflict with ssh. However, you must make sure DMLC_PS_ROOT_PORT can be seen from all containers. 

Are you using --net=host with k8s? i.e., all containers share the same host network namespace and can see the RDMA NIC?

**Comment by bobzhuyb** [comment]
In all containers, can you set DMLC_NODE_HOST to be the IP of your RDMA NIC?

**Comment by mengkai94** [comment]
hostNetwork has been used in k8s  equal to --net=host in docker.
I add setting DMLC_NODE_HOST, but does not work.
k8s using 4 physical machines also has same problem.(two woker in two physical machines, one  scheduler and one server in two cpu physical machines)

**Comment by bobzhuyb** [comment]
We can try a few things --
1. Try starting only one worker (i.e., non-distributed mode). Does the training start normally?
2. Try setting DMLC_ENABLE_RDMA=0. Does the training run normally with TCP?

Thess will help us isolate potential problems.

**Comment by mengkai94** [comment]
I try other models, and they can work. Only Resnet50 with RDMA in TF can not run. Resnet50 with TCP in TF is OK.
I find performance of VGG16 is unstable with high variance.(two woker in two physical machines, one scheduler and one server in two cpu physical machines)

**Comment by bobzhuyb** [comment]
@ymjiang Can you check the TF+ResNet50+RDMA issue? It does not make sense to me.. Maybe ResNet requires a deep queue in RDMA transport? 

@mengkai94 Regarding the VGG-16 issue, is it possible that your RDMA network configuration is not correct? Can you use BytePS's ps-lite benchmark to make sure the network works as expected? Follow the commands here https://github.com/bytedance/ps-lite/tree/byteps.  You just need to run `./tests/test_benchmark` in ps-lite folder instead of your TF script.

**Comment by ymjiang** [comment]
Looks like ResNet indeed needs large queue depth. `kReplyDepth=256` causes hanging issue when using 1 server + 2 workers.

@mengkai94 You can try one of the following to solve that:
- Use `kReplyDepth=512` if it is still a valid number on your machines;
- Use 2 servers for 2 workers so that the RDMA load will be divided and balanced.

**Comment by mengkai94** [comment]
I indeed use 2 servers for 2 workers. but I can only set  kReplyDepth=128 at maximum. so byteps can not run resnet50 in my mechine?
The results of test_benchmark are 89.6Gbps and 81Gbps in two worker with two server.
In pytorch, VGG19 is stable at 164 imgs/sec/gpu，Vgg16 can achieve 178 imgs/sec/gpu at maximum, but 145 imgs/sec/gpu on the average of 10 times.

**Comment by bobzhuyb** [comment]
@mengkai94 The kReplyDepth limitation is due to that your machine seems to be configured with lower memory registration limit. Would you check this article about how to increase the limit (if you have the root access to configure it)? https://community.mellanox.com/s/article/howto-increase-memory-size-used-by-mellanox-adapters

You can look at the example in that article and set your parameters accordingly
```
For example, if the physical memory on the server is 64GB, it is recommended to have twice this size (2x64GB=128GB) for the max_reg_mem.

max_reg_mem = (2^ log_num_mtt) * (2^1) * (4 kB)

128GB = (2^ log_num_mtt) * (2^1) * (4 kB)

2^37 = (2^ log_num_mtt) * (2^1) * (2^12)

2^24 = (2^ log_num_mtt)

24 = log_num_mtt
```

**Comment by bobzhuyb** [comment]
@mengkai94 Are you saying that the performance of PyTorch+VGG19+BytePS is stable? If so, the problem is unlikely inside ps-lite or anything below it.

**Comment by mengkai94** [comment]
Yes, the performance of PyTorch+VGG19+BytePS is stable. 
The performance of PyTorch+VGG16+BytePS:
![image](https://user-images.githubusercontent.com/27659870/76284262-c15fe780-62d7-11ea-8d8d-ccf1fc4f9919.png)

**Comment by ymjiang** [comment]
@mengkai94 The pytorch result does not look stable, either. Can you check the CPU utilization?

**Comment by ymjiang** [comment]
@mengkai94 Sorry for my misread. So the case is that VGG19 is stable while VGG16 is not. Can you also check the CPU utilization? 

BTW, what if you use just one GPU for each worker?

**Comment by zhouxhao** [comment]
> Seems like it hits the limit of memory registration. We will post a fix soon. Before that, could you please manually change the value of `kReplyDepth` (in byteps/3rdparty/ps-lite/src/rdma_utils.h) to 256 (or even smaller), and then recompile with the following commands?
> 
> ```
> cd byteps/3rdparty/ps-lite && make clean  && cd -
> pip3 uninstall -y byteps
> python3 setup.py clean --all
> python3 setup.py install
> ```

When I run run then tensoflow example synthetic_benchmark.py with ResNet50 + RDMA + two servers + two workers， it can not work. But it can work with one server and two workers.
![scheduler](https://user-images.githubusercontent.com/37922263/77733962-c0061d00-7042-11ea-8a42-244cb410c87d.jpg)

**Comment by ymjiang** [comment]
@zhouxhao Have you tried this? https://github.com/bytedance/byteps/issues/216#issuecomment-596891713

**Comment by zhouxhao** [comment]
@ymjiang

**Comment by ymjiang** [comment]
@zhouxhao Can you try to reduce `kRxDepth` only for the scheduler? Do not change it for the workers/servers if they don't run into ibv_reg_mr failed.

**Comment by bobzhuyb** [comment]
@ymjiang Maybe we should make it read from an environmental variable..

**Comment by ymjiang** [comment]
Will allow optional value in https://github.com/bytedance/ps-lite/pull/30.

**Comment by zhouxhao** [comment]
@ymjiang Thanks. It can work now. But I found it go much slower with two servers than one server. How can I find the reason?
![one](https://user-images.githubusercontent.com/37922263/77819605-02992980-7117-11ea-8195-55458b06aa73.jpg)
![two](https://user-images.githubusercontent.com/37922263/77819606-062cb080-7117-11ea-9560-cee7dabbf4f0.jpg)
![servers](https://user-images.githubusercontent.com/37922263/77819608-0927a100-7117-11ea-8f47-7bb2ce3c451b.jpg)

**Comment by ymjiang** [comment]
@zhouxhao Can you run the ps-lite benchmark to see whether the problem is in networking? You can follow the tutorial here: https://github.com/bytedance/ps-lite/tree/byteps#1-basic-benchmark

**Comment by bobzhuyb** [comment]
Closing due to inactivity.
